fix order() to handle complex subqueries

kbrock opened the pull request:
**TL;DR:** remove sql string munging for `ORDER BY` and leverage Arel instead.

### Summary

If a query has a `distinct()` and an `order()`, active record generates a pre query to just fetch ids. `columns_for_distinct` converts the order clause into something usable in the sql clause.

```ruby
Author.distinct.left_joins(:posts).where(:posts => {:id => [5,6,7,8,9,10]}).order(:name)
```

```sql
SELECT DISTINCT "authors"."id" AS alias_0, "authors"."name"
LEFT OUTER JOIN "posts" ON "posts"."author_id" = "authors"."id"
FROM "authors"
WHERE "posts"."id" in (...)
ORDER BY "posts"."name";
```

The issue enters when the `order()` clause is a subquery joining to a `has_many` model. The subquery needs to return 1 value, so it will have a `LIMIT 1` and an `ORDER BY` clause.

The sql munging in `columns_for_distinct` inadvertently modifies the subquery's `ORDER BY` clause. So the `SELECT` subquery will be different from the `ORDER BY` subquery. Postgres and MySQL do not like this:

```
ActiveRecord::StatementInvalid (PG::InvalidColumnReference: ERROR:  for SELECT DISTINCT, ORDER BY expressions must appear in select list)
LINE 1: ..."disks"."hardware_id" = "hardwares"."id" ORDER BY (SELECT  "...
```

### Before this PR

- Rails converts the `ORDER BY` clause columns to a sql string and uses regular expressions to removes  all `"ASC", "DESC"` from the sql.
- The sql in the `SELECT` no longer resembles the sql in the `ORDER BY` and the db complains.


### After

- Rails drops the `Arel::Nodes::Ordering` node and uses the actual expression (i.e.: `@expr`). Since the `Order` node is removed, the `ASC` or `DESC` are successfully removed. No String munging required.
- To support `Arel::Attributes::Attribute` nodes, the general case of `to_sql` is added.

I've observed that most of the `ORDER BY` clauses are an `Arel::Nodes::Ordering` node pointing to an `Arel::Attributes::Attribute` node. Less frequently, the `ORDER BY` clauses is an `Ordering` node pointing to a `SqlLiteral`. This supports both.

### Alternatives

- ~~I can implement `Arel::Attributes::Attribute#to_sql`, but comments suggests that `to_sql` is bad and should go away.~~ **DONE**
- ~~I can move `columns_for_distinct` into the abstract `SchemaStatements`, if you prefer.~~ **DONE**

### Thanks

Thank for all the help.

As @matthewd and @tenderlove may have guessed, this is related to [virtual attributes](https://github.com/ManageIQ/activerecord-virtual_attributes).

kbrock committed:
For distinct(order) use arel vs string munging

columns_for_distinct columns is used to make queries
work when a query has a distinct and an order clause

Before, it strictly used string and regex conversion.
Now, when available, it uses arel.

This allows subqueries that have orders in them to work

rafaelfranca commented on activerecord/lib/active_record/connection_adapters/postgresql/schema_statements.rb:
```diff
@@ -557,10 +557,25 @@ def type_to_sql(type, limit: nil, precision: nil, scale: nil, array: nil, **) #
+                  s = s.expr
+                  keep_order = true
+                end
+                # essentially s = s.to_sql but works for attributes too
```
This would be better handled using polymorphism by calling `to_sql` and it working for any class. `to_sql` should not be used by apps but rails can use it.

rafaelfranca reviewed (commented):
Is this problem specific for postgresql or can happen on mysql/sqlite too?

kbrock committed:
Introduce Attribute#to_sql

Most internal sql classes define to_sql.
This adds to_sql to Attribute

kbrock committed:
Generalize columns_for_distinct

Every database has different rules for what is a
valid order clause and whether order columns need
to be in the select column list.

But most SQL compliant databases are the same. So this just
assumes the standard implementation for mysql and postgres are
the same for other vendors

kbrock commented:
thanks @rafaelfranca - let me know if you want me to squash the to_sql into the original commit (think your rules are to NOT do that, but I thought I'd ask)

Also, you want the specs moved as well?

kbrock commented:
rebased to fix merge conflict.

rearranged commits to better partition what is done:

1. Generalize `columns_for_distinct` (and tests) per @rafaelfranca request
2. Define `to_sql` upfront.
3. Change `columns_for_distinct` implementation to rely less upon regex munging.

kamipo commented on activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb:
```diff
@@ -1148,7 +1148,26 @@ def type_to_sql(type, limit: nil, precision: nil, scale: nil, **) # :nodoc:
       #
       def columns_for_distinct(columns, orders) # :nodoc:
-        columns
+        order_columns = orders.compact_blank.map { |s|
```
I'm curious why returning order_columns for all adapters?
AFAIK this is not necessary for SQLite3.

kbrock replied:
AFAIK, this is ANSI SQL standard.

I undid for SQLite3 to be simpler.

kbrock replied:
AFAIK, this is ANSI SQL standard.

I undid for SQLite3 to be simpler.

kbrock replied:
@kamipo I have fixed to return simple query for SQLite3

Since this is typical behavior for ANSI SQL standard, I thought it better to make default behavior rather than override for mysql, postgres (and probably others)

Does this change fix your issues?

kbrock replied:
@kamipo are you ok with this pr now?

anything else that would help you understand / merge?

kbrock committed:
Do not add columns for distinct for sqlite3

kbrock reviewed (commented):
Now that SQLite3 is returned to the original, is this closer to what you want?

kbrock commented:
@rafaelfranca I moved the specs to generic connection handler.
does this make it too complicated for you - would you prefer that I make this PR smaller and just special case for mysql / postgres?
